File: Algorithm/C++/962. Maximum Width Ramp.cc
/*
 * Created on Mon Dec 24 2018 19:27:35
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// stack, O(n) time

#include <vector>
#include <stack>

using std::vector;
using std::stack;

class Solution {
 public:
  int maxWidthRamp(vector<int>& A) {
    stack<int> decreased;
    decreased.push(0);
    for (int i = 1; i < A.size(); i++) {
      if (A[i] < A[decreased.top()])
        decreased.push(i);
    }
    int result = 0;
    for (int i = A.size() - 1; i >= 0; i--) {
      while (decreased.size() > 0 && A[i] >= A[decreased.top()]) {
        result = std::max(result, i - decreased.top());
        decreased.pop();
      }
      if (decreased.size() == 0) break;
    }
    return result;
  }
};

File: Algorithm/Python/962. Maximum Width Ramp.py
# -*- coding: utf-8 -*-
# Created on Mon Dec 24 2018 14:5:43
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# stack, O(n) time
# reference: https://leetcode.com/problems/maximum-width-ramp/discuss/208348/JavaC++Python-O(N)-Using-Stack
class Solution(object):
    def maxWidthRamp(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        stack = [0]
        for i in xrange(1, len(A)):
            if A[i] < A[stack[-1]]:
                stack.append(i)
        result = 0
        for i in xrange(len(A) - 1, -1, -1):
            while stack and A[i] >= A[stack[-1]]:
                result = max(result, i - stack.pop())
            if len(stack) == 0:
                break
        return result        